// grantwell-shared/index.mjs
/**
 * Shared utilities for Node.js Lambda functions in this stack.
 *
 * Bundled as a Lambda Layer (see JsSharedLayer in functions.ts) so each
 * function can `import { ... } from 'grantwell-shared'` without duplicating
 * security-sensitive helpers across asset folders.
 */

function jsonResponse(statusCode, body) {
  return {
    statusCode,
    headers: {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  };
}

/**
 * Returns a 403 response if the API Gateway caller is not an Admin or
 * Developer. Returns null when authorization passes, or when the invocation
 * has no requestContext (e.g. EventBridge scheduled triggers, direct Lambda
 * invocations from trusted internal callers).
 */
export function requireAdmin(event) {
  if (!event?.requestContext) return null;
  try {
    const claims = event.requestContext.authorizer?.jwt?.claims || {};
    const roles = JSON.parse(claims["custom:role"] || "[]");
    if (
      !Array.isArray(roles) ||
      (!roles.includes("Admin") && !roles.includes("Developer"))
    ) {
      return jsonResponse(403, { error: "Forbidden: Admin role required" });
    }
    return null;
  } catch {
    return jsonResponse(403, { error: "Forbidden: invalid role claim" });
  }
}

/**
 * Returns the authenticated caller's Cognito sub from the API Gateway JWT
 * authorizer, or null when no requestContext is present.
 */
export function getAuthenticatedUserId(event) {
  return (
    event?.requestContext?.authorizer?.jwt?.claims?.sub ||
    event?.requestContext?.authorizer?.claims?.sub ||
    null
  );
}

// HTML sanitization for content that gets rendered through headless Chromium
// (PDF generator) or html-to-docx. Untrusted <script>/<iframe>/<style>/event
// handlers in this position would execute in the renderer with the Lambda's
// network and credential context. Strip dangerous elements with their bodies,
// drop attributes off the remaining tags, and keep only a small allowlist of
// safe formatting tags.
const ALLOWED_CONTENT_TAGS = new Set([
  "p", "br", "strong", "b", "em", "i", "u", "s", "strike",
  "ul", "ol", "li",
  "h1", "h2", "h3", "h4", "h5", "h6",
  "blockquote", "span", "div", "hr", "pre", "code", "sub", "sup",
  "table", "thead", "tbody", "tr", "td", "th",
]);

const DANGEROUS_TAGS_WITH_CONTENT = [
  "script", "style", "iframe", "object", "embed", "noscript",
  "form", "svg", "math", "template", "link", "meta", "base",
];

export function sanitizeContentHtml(input) {
  if (input == null) return "";
  let s = String(input);

  for (const tag of DANGEROUS_TAGS_WITH_CONTENT) {
    s = s.replace(new RegExp(`<${tag}\\b[\\s\\S]*?</${tag}\\s*>`, "gi"), "");
    s = s.replace(new RegExp(`<${tag}\\b[^>]*/?>`, "gi"), "");
  }

  s = s.replace(/<!--[\s\S]*?-->/g, "");

  s = s.replace(/<\/?([a-zA-Z][a-zA-Z0-9]*)\b[^>]*>/g, (match, tagName) => {
    const lower = tagName.toLowerCase();
    if (!ALLOWED_CONTENT_TAGS.has(lower)) return "";
    return match.startsWith("</") ? `</${lower}>` : `<${lower}>`;
  });

  return s;
}
